src/Navigation/Plugins/recastJSPlugin.ts

```diff
@@ -0,0 +1,319 @@
+     * @param scene is where the mesh will be added
+     * @returns debug display mesh
+     */
+    createDebugNavMesh(scene: Scene): Mesh {
+        var tri: number;
+        var pt: number;
+        var debugNavMesh = this.navMesh.getDebugNavMesh();
+        let triangleCount = debugNavMesh.getTriangleCount();
+
+        var indices = [];
+        var positions = [];
+        for (tri = 0; tri < triangleCount * 3; tri++)
+        {
+            indices.push(tri);
+        }
+        for (tri = 0; tri < triangleCount; tri++)
+        {
+            for (pt = 0; pt < 3 ; pt++)
+            {
+                let point = debugNavMesh.getTriangle(tri).getPoint(pt);
+                positions.push(point.x, point.y, point.z);
+            }
+        }
+
+        var mesh = new Mesh("NavMeshDebug", scene);
+        var vertexData = new VertexData();
+
+        vertexData.indices = indices;
+        vertexData.positions = positions;
+        vertexData.applyToMesh(mesh, false);
+        return mesh;
+    }
+
+    /**
+     * Get a navigation mesh constrained position, closest to the parameter position
+     * @param position world position
+     * @returns the closest point to position constrained by the navigation mesh
+     */
+    getClosestPoint(position: Vector3) : Vector3
+    {
+        var p = new this.bjsRECAST.Vec3(position.x, position.y, position.z);
+        var ret = this.navMesh.getClosestPoint(p);
+        var pr = new Vector3(ret.x, ret.y, ret.z);
+        return pr;
+    }
+
+    /**
+     * Get a navigation mesh constrained position, within a particular radius
+     * @param position world position
+     * @param maxRadius the maximum distance to the constrained world position
+     * @returns the closest point to position constrained by the navigation mesh
+     */
+    getRandomPointAround(position: Vector3, maxRadius: number): Vector3 {
+        var p = new this.bjsRECAST.Vec3(position.x, position.y, position.z);
+        var ret = this.navMesh.getRandomPointAround(p, maxRadius);
+        var pr = new Vector3(ret.x, ret.y, ret.z);
+        return pr;
+    }
+
+    /**
+     * Create a new Crowd so you can add agents
+     * @param maxAgents the maximum agent count in the crowd
+     * @param maxAgentRadius the maximum radius an agent can have
+     * @param scene to attach the crowd to
+     * @returns the crowd you can add agents to
+     */
+    createCrowd(maxAgents: number, maxAgentRadius: number, scene: Scene) : ICrowd
+    {
+        var crowd = new RecastJSCrowd(this, maxAgents, maxAgentRadius, scene);
+        return crowd;
+    }
+
+    /**
+     * Disposes
+     */
+    public dispose() {
+
+    }
+
+    /**
+     * If this plugin is supported
+     * @returns true if plugin is supported
+     */
+    public isSupported(): boolean {
+        return this.bjsRECAST !== undefined;
+    }
+}
+
+/**
+ * Recast detour crowd implementation
+ */
+export class RecastJSCrowd implements ICrowd {

```

Comment:
should this be exported? seem like you only want this to be created by the plugin and thus it doesn't need to be exported.

Reply:
It's used by the user to add/remove agents, tick crowd, request a GoTo.

Reply:
Maybe I'm not understanding something. Looks like to me ICrowd does those things, but the actual class doesn't need to be exposed. Are you expecting someone to new a RecastJSCrowd?

Reply:
no, crowd creation only by calling createCrowd on INavigationEnginePlugin

Reply:
Can you try not exporting this class? I think it will still work.